pkg/activator/podip_watcher.go

```diff
@@ -0,0 +1,265 @@
+	healthyDestsMux sync.RWMutex
+
+	transport http.RoundTripper
+	destsChan <-chan []string
+	logger    *zap.SugaredLogger
+}
+
+func newRevisionWatcher(destsChan <-chan []string, transport http.RoundTripper,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	rw := revisionWatcher{
+		healthyDests: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+	return &rw
+}
+
+// getHealthyDests returns a list of l4 dests which are "healthy"
+// (have returned a success on the most recent probe)
+func (rw *revisionWatcher) getHealthyDests() []string {
+	rw.healthyDestsMux.RLock()
+	defer rw.healthyDestsMux.RUnlock()
+
+	ret := []string{}
+	for dest, healthy := range rw.healthyDests {
+		if healthy {
+			ret = append(ret, dest)
+		}
+	}
+	return ret
+}
+
+func endpointsToDests(endpoints *corev1.Endpoints) []string {
+	ret := []string{}
+
+	for _, es := range endpoints.Subsets {
+		portVal := int32(0)
+		for _, port := range es.Ports {
+			if port.Name == probePortName {
+				portVal = port.Port
+				break
+			}
+		}
+
+		// This endpoint doesnt have a port we want to probe
+		// TODO(greghaynes) log error here
+		if portVal == 0 {
+			continue
+		}
+
+		for _, addr := range es.Addresses {
+			if addr.Hostname != "" {
+				ret = append(ret, fmt.Sprintf("%s:%d", addr.Hostname, portVal))
+			} else {
+				ret = append(ret, fmt.Sprintf("%s:%d", addr.IP, portVal))
+			}
+		}
+	}
+
+	return ret
+}
+
+func (rw *revisionWatcher) checkDests() {
+	healthyDests := make(map[string]bool)
+	var probeWg sync.WaitGroup
+
+	for _, dest := range rw.dests {
+		// If the dest is already healthy then save this
+		if curHealthy, ok := rw.healthyDests[dest]; ok && curHealthy {
+			healthyDests[dest] = true
+		} else {
+			probeWg.Add(1)
+			go func(dest string) {
+				httpDest := url.URL{
+					Scheme: "http",
+					Host:   dest,
+				}
+				ok, err := prober.Do(context.Background(), rw.transport, httpDest.String(),
+					prober.WithHeader(network.ProbeHeaderName, queue.Name),
+					prober.ExpectsBody(queue.Name))
+				healthyDests[dest] = ok
+
+				if err != nil {
+					rw.logger.Errorw(fmt.Sprintf("Probing of dest %q failed", dest), zap.Error(err))
+				} else if !ok {
+					rw.logger.Infof("Probing of dest %q failed.", dest)
+				}
+
+				probeWg.Done()
+			}(dest)
+		}
+	}
+
+	probeWg.Wait()
+
+	rw.healthyDestsMux.Lock()
+	defer rw.healthyDestsMux.Unlock()
+	rw.healthyDests = healthyDests
+}
+
+func (rw *revisionWatcher) runOnce(tickCh <-chan time.Time) bool {
+	select {
+	case dests := <-rw.destsChan:
+		if dests == nil {
+			// shutdown
+			return true
+		}
+		rw.dests = dests
+		rw.checkDests()
+	case <-tickCh:
+		rw.checkDests()
+	}
+	return false
+}
+
+func (rw *revisionWatcher) run() {
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		if rw.runOnce(ticker.C) {
+			return
+		}
+	}
+}
+
+// PodIPWatcher listens to revision endpoints and keeps track of healthy
```

Comment:
So this is across revision watcher? Probably should be more obviously articulated?